pkg/ingester/instance.go

```diff
@@ -218,27 +233,40 @@ func (i *instance) Push(ctx context.Context, req *logproto.PushRequest) error {
 		}
 	}
 
-	return appendErr
+	return mountPushGRPCError(appendErr, statusCode, details)
 }
 
-// errorForFailedEntries mounts an error to be returned in a GRPC call based on entries that couldn't be ingested.
+// mountPushGRPCError returns an enriched gRPC error or nil if no error is given.
 //
-// The returned error is enriched with the gRPC error status and with a list of details.
-// As of now, the list of details is a list of all streams that were rate-limited. This list can be used
-// by the distributor to fine tune streams that were limited.
-func errorForFailedEntries(s *stream, failedEntriesWithError []entryWithError, totalEntries int) error {
-	if len(failedEntriesWithError) == 0 {
+// This is in its own function to be reused by tests.
+func mountPushGRPCError(err error, statusCode spb.Code, details []*types.Any) error {
+	if err == nil {
 		return nil
 	}
 
+	return status.ErrorProto(&spb.Status{
+		Code:    int32(statusCode),
+		Message: err.Error(),
+		Details: details,
+	})
+}
+
+// errorForFailedEntries mounts an error message and a status code based on entries that couldn't be ingested.
+//
+// The returned error has an error line for every entry that failed to be ingested.
+func errorForFailedEntries(s *stream, failedEntriesWithError []entryWithError, totalEntries int) (spb.Code, error) {
+	if len(failedEntriesWithError) == 0 {
+		return 0, nil
+	}
+
 	lastEntryWithErr := failedEntriesWithError[len(failedEntriesWithError)-1]
 	_, ok := lastEntryWithErr.e.(*validation.ErrStreamRateLimit)
 	outOfOrder := chunkenc.IsOutOfOrderErr(lastEntryWithErr.e)
 	if !outOfOrder && !ok {
-		return lastEntryWithErr.e
+		return 0, lastEntryWithErr.e
```

Comment:
If we're using gRPC codes, this means `OK`. Looking at the `httpgrpc.Errorf` and `httpgrpc.HTTPResponseFromError` code, they aren't doing any special code conversions so we can just stick with http codes and use `http.BadRequest`

Reply:
Oh shit, good catch. `0` here is definitely a mistake: I think I was previously returning as the first argument the number of entries that were ingested or something, and probably overlooked this part of the code :smile: